app/multitenant/dynamo_collector.go

```diff
@@ -345,6 +408,72 @@ func (c *dynamoDBCollector) getReports(userid string, row int64, start, end time
 	return append(cachedReports, fetchedReport...), nil
 }
 
+func memcacheStatusCode(err error) string {
+	// See https://godoc.org/github.com/bradfitz/gomemcache/memcache#pkg-variables
+	switch err {
+	case nil:
+		return "200"
+	case memcache.ErrCacheMiss:
+		return "404"
+	case memcache.ErrMalformedKey:
+		return "400"
+	default:
+		return "500"
+	}
+}
+
+func (c *dynamoDBCollector) fetchFromMemcache(reportKeys []string) ([]report.Report, []string, error) {
+	var foundReports map[string]*memcache.Item
+	err := timeRequest("Get", memcacheRequestDuration, memcacheStatusCode, func() error {
+		var err error
+		foundReports, err = c.memcache.GetMulti(reportKeys)
+		if err != nil {
+			return err
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, reportKeys, err
+	}
+
+	// Decode all the reports in parallel.
+	type result struct {
+		key    string
+		report *report.Report
+	}
+	ch := make(chan result, len(reportKeys))
```

Comment:
I would be tempted to have two channels, one for missing keys and one for reports, and unify the error handling and the missing handling in the loop below, something like:

```
missing := make(chan string, len(...))
reports := make(chan report, len(...))
for _, key := range reportKeys {
    item, ok := foundReports[key]
    go func(key string, buf []byte, bool found) {
        if !ok {
            missing <- key
            return
        }        
        
        rep, err := report.MakeFromBinary(bytes.NewReader(item.Value))
        if err != nil {
            log...
            missing <- key
            return
        }
        
        reports <- rep
    }(reportKey, item, ok)
}
```

WDYT?